// react-native-quick-crypto/lib/commonjs/NativeQuickCrypto/Cipher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KeyVariantLookup = exports.KeyVariant = void 0;
// TODO: until shared, keep in sync with C++ side (cpp/Utils/MGLUtils.h)
let KeyVariant = exports.KeyVariant = /*#__PURE__*/function (KeyVariant) {
  KeyVariant[KeyVariant["RSA_SSA_PKCS1_v1_5"] = 0] = "RSA_SSA_PKCS1_v1_5";
  KeyVariant[KeyVariant["RSA_PSS"] = 1] = "RSA_PSS";
  KeyVariant[KeyVariant["RSA_OAEP"] = 2] = "RSA_OAEP";
  KeyVariant[KeyVariant["DSA"] = 3] = "DSA";
  KeyVariant[KeyVariant["EC"] = 4] = "EC";
  KeyVariant[KeyVariant["NID"] = 5] = "NID";
  KeyVariant[KeyVariant["DH"] = 6] = "DH";
  return KeyVariant;
}({});
const KeyVariantLookup = exports.KeyVariantLookup = {
  'RSASSA-PKCS1-v1_5': KeyVariant.RSA_SSA_PKCS1_v1_5,
  'RSA-PSS': KeyVariant.RSA_PSS,
  'RSA-OAEP': KeyVariant.RSA_OAEP,
  ECDSA: KeyVariant.DSA,
  ECDH: KeyVariant.EC,
  Ed25519: KeyVariant.NID,
  Ed448: KeyVariant.NID,
  X25519: KeyVariant.NID,
  X448: KeyVariant.NID,
  DH: KeyVariant.DH
};
